File: LeetCode/python/609.py
class Solution(object):
    def findDuplicate(self, paths):
        """
        :type paths: List[str]
        :rtype: List[List[str]]
        """
        duplicated = {}
        for path in paths:
            str = path.split(' ')
            folder = str[0] + '/'
            for i in range(1, len(str)):
                parentheses = str[i].find('(')
                content = str[i][parentheses + 1:str[i].find(')')]
                file = str[i][:parentheses]
                if content not in duplicated:
                    duplicated[content] = []
                duplicated[content].append(folder + file)
        ret = []
        for content, files in duplicated.items():
            if len(files) > 1:
                ret.append(files)
        return ret

File: LeetCode/cpp/609.cpp
class Solution {
public:
    vector<vector<string>> findDuplicate(vector<string> &paths) {
        vector<vector<string>> ret;
        unordered_map<string, vector<string>> duplicate_files;
        for (auto &s:paths) {
            s += ' ';
            int start = 0;
            int space = s.find(' ', start);
            string path = s.substr(0, space) + '/';
            start = space + 1;
            while (s.find(' ', start) != string::npos) {
                string str = s.substr(start, s.find(' ', start) - start);
                int parentheses = str.find('(');
                string content = str.substr(parentheses + 1, str.find(')') - parentheses - 1);
                string file = path + str.substr(0, parentheses);
                if (duplicate_files.find(content) == duplicate_files.end())
                    duplicate_files.insert(pair<string, vector<string>>(content, vector<string>()));
                duplicate_files[content].push_back(file);
                start += str.size() + 1;
            }
        }
        for (auto &iter:duplicate_files) {
            if (iter.second.size() > 1)
                ret.push_back(iter.second);
        }
        return ret;
    }
};
